Introducing Object Cores

peter-mitsis opened the pull request:
Introducing object cores in response to #57672.

Zephyr objects can have a special structure embedded within them (an object core). Zephyr objects of the same type are linked together through this object core. Additional functionality can be linked to the object core (such as statistics tracking). This can then provide a uniform means to access that additional functionality without needing to develop many custom interfaces. In the case of object cores, debugging tools can walk the object lists and obtain information about the objects.

The current iteration of object cores is quite minimal. The have a list node, a back pointer to the object type list of which it is a part, and an optional pointer to "raw" stats information. The object type contains a list node (linking different object types together), a list of all linked zephyr objects of that type, an ID and obj_core offset (to be leveraged by debugging tools to be able to identify and easily find relevant data of the object type), and an optional pointer to the a statistics descriptor.

Basic statistics operations include getting raw stats data, cooked stats data, resetting, enabling and disabling stats gathering. Not all functionality is required to be implemented--only what makes sense for the object.

peter-mitsis committed:
kernel: Add initial obj_core infrastructure

Adds the initial object core infrastructure.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Re-factor k_mem_slab definition

Rearranges the k_mem_slab fields so that information that describes
how much of the memory slab is used is co-located. This will allow
easier of its statistics into the object core statistics reporting
framework.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Integrate object cores into kernel

Integrates object cores into the following kernel structures
   sys_mem_blocks, k_mem_slab
   _cpu, z_kernel
   k_thread, k_timer
   k_condvar, k_event, k_mutex, k_sem
   k_mbox, k_msgq, k_pipe, k_fifo, k_lifo, k_stack

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Re-factor sys_mem_blocks definition

Rearranges the sys_mem_blocks fields so that information that describes
how much of the memory block is used is co-located. This will allow
easier of its statistics into the object core statistics reporting
framework.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Refactor CPU usage

Refactors CPU usage (thread runtime stats) to make it easier to
integrate with the object core statistics framework.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
test: Implement object core test

Creates a test to verify that kernel objects can be found within
the object core framework.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: FIFO and LIFO have their own sections

Instead of placing both FIFOs and LIFOs into a common k_queue
type section, they are now placed into their own custom iterable
secitons. This is necessary when employing object cores as FIFOs
and LIFOs will be associated with different lists.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Integrate object core statistics

Integrates object core statistics framework into the following
kernel objects:
  sys_mem_blocks, k_mem_slab
  threads, _cpu, z_kernel

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
test: Implement object core stats API test

Creates a set of tests to verify that the top-level
k_obj_core_stats_xxx() routines behave as expected.
Note that this test is not meant to test the details
of the object core statistics operator function pointers.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
test: Implement obj core stats test

Adds a test to test the objects integrated into the object core
statistics framework.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

nashif commented:
documentation? :-)

peter-mitsis commented:
Documentation? Gah! I knew I had to be forgetting something.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+#define OBJ_TYPE_SEM_ID        0x53454d34   /* "SEM4" in ASCII */
+#define OBJ_TYPE_STACK_ID      0x5354434b   /* "STCK" in ASCII */
+#define OBJ_TYPE_THREAD_ID     0x54485244   /* "THRD" in ASCII */
+#define OBJ_TYPE_TIMER_ID      0x54494d52   /* "TIMR" in ASCII */
```
This is cute and all, but if you're going to go through the trouble, I demand that this computation be performatively elegant and not manual:

```c
#define KOBJ_TYPE_ID_NONSENSE(s) (s[0]<<24)|(s[1]<<16)|(s[2]<<8)|(s[3])
#define OBJ_TYPE_CONDVAR_ID KOBJ_TYPE_ID_NONSENSE("COND")
```

Also, I note that these values are all expressed... big-endian?!

andyross replied:
Seriously but pedantically: "OBJ_" probably isn't a unique enough prefix and we need a K_ on there too (or just a double underscore if I'm understanding the current direction of our style guide?).  This is a /include header that can find its way into user code, so everything needs to be namespaced, even "internal" values.

peter-mitsis replied:
Good points. Will update.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+
+struct k_obj_core {
+	sys_dnode_t        node;   /* Object node within object type's list */
+	struct k_obj_type *type;   /* Object type to which object belongs */
```
As specified, this is 12 bytes for every "object" in the kernel.  That's relatively heavyweight.  Maybe use an slist instead?  Is O(N) deletion speed really a problem in tracing contexts?  In general Zephyr does very little "deletion" of kernel things.

peter-mitsis replied:
Can do.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+struct k_obj_type {
+	sys_dnode_t    node;   /* Node within list of object types */
+	sys_dlist_t    list;   /* List of objects of this object type */
+	uint32_t       id;     /* Unique type ID */
```
These are those type IDs from earlier.  But you have a struct here too, and are using the struct address to idenfity the type.  Why do we need the ID?

peter-mitsis replied:
The ID exists for tools to be able to easily identify the object type strictly by looking at the memory contents. Essentially, I am imagining a Zephyr aware debugger that is looking at memory contents and be able to properly identify the type (and associated objects). Otherwise, how could it "know" whether the object type is for threads or memory blocks or something else entirely.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+ * @param id A means to identify the object type
+ * @param off Offset of object core within the structure
+ */
+extern void z_obj_type_init(struct k_obj_type *type, uint32_t id, size_t off);
```
For completeness, might be best to return the struct k_obj_type pointer instead of void.  Any further use of the type is going to need that anyway.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+ * does not prevent the addition or removal of object cores while walking the
+ * list. Although this is expected to offer better responsiveness, it may
+ * result in unexpected behavior. Note that locks are enabled while invoking
+ * the callback.
```
Same comment: what exactly does "locks are enabled" mean?

peter-mitsis replied:
I will need to clarify.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+ * This function copies the raw statistics associated with the kernel object
+ * core specified by @a obj_core into the buffer @a stats. Note that the size
+ * of the buffer (@a stats_len) must match the size specified by the kernel
+ * object type's statistics descriptor.
```
As above: What statistics descriptor? :)  The API for that is missing in this patch.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+ * @brief Retrieve the statistics associated with the kernel object
+ *
+ * This function copies the statistics (not raw) associated with the kernel
+ * object core specified by @a obj_core into the buffer @a stats. Note that the
```
I'm confused.  What's the difference between a "raw" statistics buffer and a "not raw" statistics buffer?

peter-mitsis replied:
I probably should have used "cooked" instead of "not raw". Raw = raw data. Cooked = operations may have been done to the data to make it more palatable for human consumption (calculating averages for example).

andyross commented on kernel/Kconfig:
```diff
@@ -436,6 +436,113 @@ menuconfig OBJ_CORE
+	  core framework.
+
+config OBJ_CORE_FIFO
+	bool "Integrate FIFOs into object core framework"
```
Seems odd to have fifos/lifos share all the same implementation except... tracking and stats?  Shouldn't you just add the obj_core to k_queue instead?

peter-mitsis replied:
That was the original plan until I realised that interferes with the K_OBJ_CORE() macro. That is, the caller to K_OBJ_CORE() when using fifo/lifo's would have to pass a pointer to the underlying k_queue (somewhat exposing implementation details), or we would need a K_FIFO_OBJ_CORE() / K_LIFO_OBJ_CORE(). I wasn't particularly fond of either of those options.

andyross reviewed (commented):
Some notes.  Mostly this seems reasonable.  It would definitely good to see the patch ordering cleaned up for bisectability reasons.

I will say that I'm a little confused about how the stats framework is intended to work.  Mostly I'd imagine "stats" to look like a per-type struct with fields that the type uses directly, like an set of atomic counts or a "last event time" value or whatever.  And each type would then expose those stats for analysis via a "here's my stats" API of its own.  Instead it looks like everyone is supposed to copy all the stats at once for every object in an opaque blob?  What's the advantage there?

peter-mitsis commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+ * Tools may use this list as an entry point to identify all registered
+ * object types and the object cores linked to them.
+ */
+extern sys_dlist_t obj_type_list;
```
Naming kills me. Happy to do this.

peter-mitsis commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+ * @retval NULL if object type not found
+ * @retval Pointer to object type if found
+ */
+extern struct k_obj_type *k_obj_type_find(uint32_t type_id);
```
I can be down with that. :)

peter-mitsis commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,283 @@
+static inline void k_obj_type_stats_init(struct k_obj_type *type,
+					 struct k_obj_stats_desc *stats_desc)
+{
+	type->stats_desc = stats_desc;
```
There was a lot of squashing and reordering going on. The occasional item has slipped through.

peter-mitsis committed:
kernel: Add initial k_obj_core_stats infrastructure

Adds the infrastructure to integrate statistics into
the object core.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
doc: Add object core documentation

Adds documentation for both object cores and object core statistics.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -4751,6 +4790,9 @@ extern void k_mbox_data_get(struct k_mbox_msg *rx_msg, void *buffer);
 
 /** Pipe Structure */
 struct k_pipe {
+#ifdef CONFIG_OBJ_CORE_PIPE
```
Good point. I will fix that.

peter-mitsis replied:
That should be fixed now.

peter-mitsis commented on kernel/obj_core.c:
```diff
@@ -0,0 +1,350 @@
+
+		/* Object type not configured for statistics. */
+
+		k_spin_unlock(&lock, key);
```
Generally, it is not expected that things will change. The locking is done as more of a defensive practice to ensure that no funny business occurs there if a developer decides to get fancy/clever with their own custom application of the obj_core framework on their own objects.

peter-mitsis commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,340 @@
+ *
+ * @return 0 on success, -errno on failure
+ */
+extern int k_obj_core_stats_disable(struct k_obj_core *obj_core);
```
Registering/deregistering is done to interface some statistics functionality with k_obj_core_stats_xxx() APIs. The stats gathering can still be done independently of this (such as thread runtime stats).
k_obj_core_stats-[en/dis]able() allows the stats gathering to be enabled/disabled using the k_obj_core_stats_xxx() interface.

edersondisouza commented on lib/os/mem_blocks.c:
```diff
@@ -16,37 +16,40 @@ static void *alloc_blocks(sys_mem_blocks_t *mem_block, size_t num_blocks)
-	r = sys_bitarray_alloc(mem_block->bitmap, num_blocks, &offset);
-	if (r == 0) {
-
+	r = sys_bitarray_alloc(mem_block->bitmap, info.num_blocks, &offset);
```
This `info.num_blocks` change shouldn't be here - it is fixed down the line in another patch, but things would be simpler without this extraneous change.

peter-mitsis replied:
Fixing that ...

edersondisouza commented on lib/os/mem_blocks.c:
```diff
@@ -454,3 +455,29 @@ int sys_mem_blocks_runtime_stats_reset_max(sys_mem_blocks_t *mem_block)
+					 sys_mem_blocks *, block_pp) {
+		k_obj_core_init(K_OBJ_CORE(*block_pp),
+				&obj_type_sys_mem_blocks);
+		k_obj_core_link(K_OBJ_CORE(*block_pp));
```
All these pairs `k_obj_core_init` and `k_obj_core_link` makes me wonder why not a `k_obj_core_init_and_link` one...

peter-mitsis replied:
Seems reasonable.

edersondisouza commented on lib/os/mem_blocks.c:
```diff
@@ -364,15 +365,15 @@ int sys_multi_mem_blocks_alloc(sys_multi_mem_blocks_t *group,
 
 	if ((ret == 0) && (blk_size != NULL)) {
-		*blk_size = BIT(allocator->blk_sz_shift);
+		*blk_size = BIT(allocator->info.blk_sz_shift);
```
I guess this and previous changes actually belong to the "refactor sys_mem_blocks" patch.

peter-mitsis replied:
The perils of continuous refactoring ... Fixing.

edersondisouza commented on kernel/obj_core.c:
```diff
@@ -112,3 +115,236 @@ int k_obj_type_walk_unlocked(struct k_obj_type *type,
 }
+
+#ifdef CONFIG_OBJ_CORE_STATS
+int k_obj_stats_register(struct k_obj_core *obj_core, void *stats,
```
I guess this and following functions are called `k_obj_core_stats...` instead of `k_obj_stats...`, from the `obj_core.h` above.

peter-mitsis replied:
Fixing.

edersondisouza commented on kernel/obj_core.c:
```diff
@@ -314,10 +314,10 @@ int k_obj_stats_disable(struct k_obj_core *obj_core)
 {
 	int  rv;
-	struct k_obj_stats_desc *desc;
+	struct k_obj_core_stats_desc *desc;
```
Oh, and here they are fixed. Maybe just squash the changes in this file to the previous patch?

peter-mitsis replied:
Will do.

nashif commented:
@peter-mitsis  did you look into the failing test?

peter-mitsis commented:
I thought I had previously, and (previously) thought that it was failing on an unrelated test. But now I see it failing on one of the ones that I added to test the feature.  Looking into this further ....

peter-mitsis commented:
The failing test gives the following message.

ERROR: Test 'test_obj_core_stats_deregister' assigned to test suite 'obj_core_stats_api' which doesn't exist

This is an unusual one to me.

peter-mitsis commented:
Hmmm, I am unable to reproduce this failure locally. :(

peter-mitsis commented:
After some additional digging, I suspect that this failing test is due to a compiler bug in the ARC64 toolchain used in Zephyr SDK 0.16.2.

Debugging is tricky because depending where the debug code is added, we can get either a pass or a failure. Here's what I did and what I observed ...

Update ztest routine ztest_find_test_suite() to print both strings and their addresses since the failure message indicates that this routine is not finding the expected suite name. It fails on the first iteration of the for-loop in ztest_find_test_suite(). The address of the string is observed to match that of _ztest_unit_test_list_end. This is particularly suspicious since the for-loop that invokes ztest_find_test_suite() uses _ztest_unit_test_list_end as a comparison to terminate the loop. Furthermore, the generated assembly, if I am deciphering it correctly, is placing that address in register r0 prior to calling ztest_find_test_suite(). The relevant code is pasted below.

```
    804020c4:   5e0a 1f80 8040 74b8     movl    r14,0x804074b8 # _ztest_unit_test_list_start
    804020cc:   580a 0f80 8040 75d0     movl    r0,0x804075d0  # _ztest_unit_test_list_end
    804020d4:   0e53 100c               brlol   r14,r0,82       ;80402126  # BODY of 2nd for-loop
...
# BODY of 2nd for-loop  (Note r0 has the wrong address in it).
    80402126:   09a6 ffef               bl.d    -1628   ;80401ac8 <ztest_find_test_suite>
    8040212a:   1600 1840               ldl     r0,[r14]
```
I have also observed that adding something like printk() before the call ztest_find_test_suite() results in changed generated assembly that ends up putting the correct value into r0, which (again if I am deciphering this code correctly) is used to pass the parameter to the routine.

peter-mitsis committed:
test: Shorten object core API stats test name

Changing the name of the routine test_obj_core_stats_deregister()
to something shorter (test_obj_core_stats_dereg) has been found to
be a workaround to the issue identified in #61087.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis commented:
This CI failure is not related to the proposed changes. The exact same failure can be replicated without them.

nashif commented:
@peter-mitsis  the new ci failure looks real and related to your change

> This CI failure is not related to the proposed changes. The exact same failure can be replicated without them.

yes, the current one is not related.

peter-mitsis commented:
Updated with a newer version to resolve conflicts due to recent k_mem_slab_free() API changes in the main branch.

peter-mitsis commented:
Gah! More failures unrelated to the proposed changes. :(

nashif commented:
@andyross  can you please take another look?

andyross commented on kernel/events.c:
```diff
@@ -336,3 +344,24 @@ uint32_t z_vrfy_k_event_wait_all(struct k_event *event, uint32_t events,
 #endif
+
+#ifdef CONFIG_OBJ_CORE_EVENT
+static int init_event_module(void)
```
Style: the term "module" is pretty heavily overloaded, and this doesn't really represent a "module" abstraction in the way the word gets used anywhere else.  It's a 100% local name, would be clearer as "init_obj_core_list()" or whatever as that's all it's actually doing.

peter-mitsis replied:
Can do.

andyross commented on kernel/obj_core.c:
```diff
@@ -0,0 +1,357 @@
+	int  rv;
+	struct k_obj_core_stats_desc *desc;
+
+	__ASSERT((obj_core != NULL), "NULL parameter");
```
Style: all this checking for NULL seems wasteful to me.  That's (1) a clear and obvious software bug in the app or subsystem and (2) really obvious in a debugger when it happens at runtime.  I don't think the assertion is providing much value, and promiscuous assertions tend over the long term to **discourage** people from using asserts  (because the app doesn't fit when it's enabled!).   Zephyr CONFIG_ASSERT has always been comparatively lightweight, this doesn't look lightweight to me.

peter-mitsis replied:
Good point. Will adjust.

andyross commented on kernel/mailbox.c:
```diff
@@ -447,3 +455,25 @@ int k_mbox_get(struct k_mbox *mbox, struct k_mbox_msg *rx_msg, void *buffer,
+
+	STRUCT_SECTION_FOREACH(k_mbox, mbox) {
+		k_obj_core_init_and_link(K_OBJ_CORE(mbox), &obj_type_mailbox);
+	}
```
I had another comment on one of these but can't find it.  I'm realizing that.

1. This **exact** init code is being repeated verbatim for every object type, the only difference being the type-specific stuff (i.e. the pointer to the type object, the type ID, the name-of the section).

2. There's no actual type-specific logic being implemented in any of them, at all.  They're all just initializing the type struct and constructing a runtime list of the static objects.

3. All that info is known a-priori at build time already.

That's a ton of boiler plate code and duplicated instruction bytes.  Can't you construct a table at build time such that a single initialization pass can be done from just one function?

peter-mitsis replied:
This is a good idea. I will create an enhancement so this does not get lost.

andyross commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,353 @@
+ * the list without obtaining the global spinlock. Consequently, this routine
+ * does not prevent the addition or removal of object cores while walking the
+ * list. Although this is expected to offer better responsiveness, it may
+ * result in unexpected behavior.
```
FWIW I'd be stronger with the prescription here, e.g. "No synchronization is provided here.  Mutation of the list of objects while this function is in-progress must be prevented at the application layer, otherwise undefined/unreliable behavior, corruption, and crashes may result."

peter-mitsis replied:
I like your phrasing. It shall be so. :)

andyross reviewed (dismissed):
Haven't looked at this in a while.  My concerns earlier were addressed.  I don't think it rises to a -1, but I will say the amount of initialization boilerplate in the final version bugs me pretty badly.  That's not very Zephyrish; we've always preferred generation to repetition and build-time to run-time overhead.

jukkar reviewed (dismissed):
I have done a small experiment with this to get information about the network sockets in the system. Did not see any issues and things worked well. I will send a PR after this one is merged.

jhedberg commented on doc/kernel/object_cores/index.rst:
```diff
@@ -0,0 +1,234 @@
+    void my_obj_type_init(void)
+    {
+        z_obj_type_init(&my_obj_type, K_OBJ_TYPE_MY_NEW_TYPE,
+                        offsetof(struct my_new_type, obj_core);
```
Shouldn't this be `struct my_new_obj` instead of `struct my_new_type`? I.e. what you have on line 93.

kartben replied:
@peter-mitsis would it make things easier to add an actual code sample demonstrating this? This would help keep the code free of typos, and you wwould still be embed to embed snippets of it in this documentation page using `.. literalinclude:` if you wanted to.

peter-mitsis replied:
@jhedberg ... you are correct. I will be fixing this shortly.

@kartben ... I am working on a sample project for this.

kartben commented on doc/kernel/object_cores/index.rst:
```diff
@@ -0,0 +1,234 @@
+linked together via a singly linked list. Together, this can allow debugging
+tools to traverse all the objects in the system.
+
+Object cores have been integrated into following kernel objects:
+ * Condition Variables
+ * Events
+ * FIFOs and LIFOs
+ * Mailboxes
+ * Memory Slabs
+ * Message Queues
+ * Mutexes
+ * Pipes
+ * Semaphores
+ * Timers
+ * System Memory Blocks
```
the leading white spaces is causing the list to render as a definition list, so you need to drop it. Also would be nice to link to doc pages (I've added explicit aliases for the pages to keep the list easy to read when browsing the source.

```suggestion
Object cores have been integrated into following kernel objects:

* :ref:`Condition Variables <condvar>`
* :ref:`Events <events>`
* :ref:`FIFOs <fifos_v2>` and ref:`LIFOs <lifos_v2>`
* :ref:`Mailboxes <mailboxes_v2>`
* :ref:`Memory Slabs <memory_slabs_v2>`
* :ref:`Message Queues <message_queues_v2>`
* :ref:`Mutexes <mutexes_v2>`
* :ref:`Pipes <pipes_v2>`
* :ref:`Semaphores <semaphores_v2>`
* :ref:`Timers <timers_v2>`
* :ref:`System Memory Blocks <sys_mem_blocks>`
```

peter-mitsis replied:
Fixing.

kartben replied:
> Fixing.

thx! If you have a chance please test locally: I might have gotten some references wrong as i provided my feedback through GitHub UI and did not actually test this change

kartben commented on doc/kernel/object_cores/index.rst:
```diff
@@ -0,0 +1,234 @@
+statistics. Queried data may be "raw", but it may also have been manipulated in
+some way by calculation (such as determining an average).
+
+Object core statistics have been integrated into the following kernel objects:
```
typo / copy-paste error?

```suggestion
```

peter-mitsis replied:
Fixing.

kartben commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,353 @@
+};
+
+struct k_obj_type {
+	sys_snode_t    node;   /* Node within list of object types */
```
either make the field comments javadoc-style to expose documentation, or hide them if they are internal.

peter-mitsis replied:
Fixing.

kartben commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,353 @@
+#endif
+};
+
+struct k_obj_core {
+	sys_snode_t        node;   /* Object node within object type's list */
```
ditto

peter-mitsis replied:
Fixing.

kartben commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,353 @@
+};
+
+/**
+ * @defgroup obj_core_apis Object Core APIs
```
Don't you want to move this further up? `k_obj_type`, `k_obj_core`, ... are API, I believe? Note how right now you don't have  them showing up in the index.rst page you created.

Also make sure to actually document everything that's API in this file, and hide what isn't behind `@cond INTERNAL_HIDDEN`.

kartben commented on doc/kernel/object_cores/index.rst:
```diff
@@ -0,0 +1,234 @@
+Object Core Concepts
+********************
+
+Each instance of an object has embedded within it an object core field named
```
nit:

```suggestion
Each instance of an object embeds an object core field named
```

peter-mitsis replied:
Fixing.

kartben commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,353 @@
+ * @param stats Pointer to raw kernel statistics
+ * @param stats_len Size of raw kernel statistics buffer
+ *
+ * @return 0 on success, -errno on failure
```
nit: there is an inconsistent use of `@return` / `@retval` throughout the file

peter-mitsis replied:
Fixing the inconsistency. I still have to learn more about the 'literalinclude' before I am comfortable with it.

kartben reviewed (changes requested):
Made a quick pass on the documentation (both doxygen and main  index.rst page)

kartben commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,353 @@
+#endif
+};
+
+struct k_obj_core {
```
```suggestion
/** Object core structure */
struct k_obj_core {
```

peter-mitsis replied:
Fixing.

peter-mitsis commented on include/zephyr/kernel/obj_core.h:
```diff
@@ -0,0 +1,353 @@
+	int (*disable)(struct k_obj_core *obj_core);
+	int (*enable)(struct k_obj_core *obj_core);
+};
+
```
Fixing

peter-mitsis commented:
@kartben - This set of changes should address the documentation issues you previously raised.

kartben reviewed (dismissed):
> @kartben - This set of changes should address the documentation issues you previously raised.

Awesome! +1 from a doc perspective.

peter-mitsis commented:
Needed to fix a build issue stemming from the use of z_current_get() instead of k_current_get().  (A recent commit renamed z_current_get).

teburd reviewed (dismissed):
What a neat feature. The docs page does an excellent job explaining the idea and layout.

PR looks great.

peter-mitsis committed:
west.yml: Pull in latest percepio updates

Pulls in the latest percepio module updates that alter how it
accesses the k_mem_slab structure (due to changes from this PR).

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

nashif commented:
@jhedberg  andyross already approved that earlier, please take this into 3.5 :)

kartben reviewed (commented):
+1 for docs
